src/vs/workbench/contrib/search/browser/notebookSearch/notebookSearchService.ts

```diff
@@ -0,0 +1,212 @@
+
+			const experimentalNotebooksEnabled = this.configurationService.getValue<ISearchConfigurationProperties>('search').experimental?.closedNotebookRichContentResults ?? false;
+
+			let closedResultsPromise: Promise<IClosedNotebookSearchResults | undefined> = Promise.resolve(undefined);
+			if (experimentalNotebooksEnabled) {
+				closedResultsPromise = this.getClosedNotebookResults(query, new ResourceSet(localNotebookFiles, uri => this.uriIdentityService.extUri.getComparisonKey(uri)), token ?? CancellationToken.None);
+			}
+
+			const promise = Promise.all([localResultPromise, closedResultsPromise]);
+			return {
+				completeData: promise.then((resolvedPromise) => {
+					const openNotebookResult = resolvedPromise[0];
+					const closedNotebookResult = resolvedPromise[1];
+
+					const resolved = resolvedPromise.filter((e): e is IOpenNotebookSearchResults | IClosedNotebookSearchResults => !!e);
+					const resultArray = [...openNotebookResult.results.values(), ...closedNotebookResult?.results.values() ?? []];
+					const results = arrays.coalesce(resultArray);
+					if (onProgress) {
+						results.forEach(onProgress);
+					}
+					this.logService.trace(`local notebook search time | ${searchLocalEnd - searchStart}ms`);
+					return <ISearchComplete>{
+						messages: [],
+						limitHit: resolved.reduce((prev, cur) => prev || cur.limitHit, false),
+						results,
+					};
+				}),
+				allScannedFiles: promise.then(resolvedPromise => {
+					const openNotebookResults = resolvedPromise[0];
+					const closedNotebookResults = resolvedPromise[1];
+					const results = arrays.coalesce([...openNotebookResults.results.keys(), ...closedNotebookResults?.results.keys() ?? []]);
+					return new ResourceSet(results, uri => this.uriIdentityService.extUri.getComparisonKey(uri));
+				})
+			};
+		};
+		const promiseResults = getAllResults();
+		return {
+			openFilesToScan: new ResourceSet(localNotebookFiles),
+			completeData: promiseResults.completeData,
+			allScannedFiles: promiseResults.allScannedFiles
+		};
+	}
+
+	private async getClosedNotebookResults(textQuery: ITextQuery, scannedFiles: ResourceSet, token: CancellationToken): Promise<IClosedNotebookSearchResults> {
+
+		const sortedContributedTypes = [...this.notebookService.getContributedNotebookTypes()].sort((a, b) => priorityToRank(b.priority) - priorityToRank(a.priority));
+
+		const getResultsFromProviderInfo = async (providerInfo: NotebookProviderInfo) => {
+			const serializer = (await this.notebookService.withNotebookDataProvider(providerInfo.id)).serializer;
+			return await serializer.searchInNotebooks(textQuery, token);
+		};
+
+		const start = Date.now();
+		const searchComplete = (await Promise.all(sortedContributedTypes.map(async e => await getResultsFromProviderInfo(e))));
+		const results = searchComplete.flatMap(e => e.results);
+		let limitHit = searchComplete.some(e => e.limitHit);
+
+		// results are already sorted with high priority first, filter out duplicates.
```

Comment:
this is to ensure that files matched by multiple serializers only gets results from one.